cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/DefaultMetadataServiceClientTest.java

```diff
@@ -0,0 +1,139 @@
+    new ScopedNameOfKind(MetadataKind.TAG, MetadataScope.USER, "c"),
+    MetadataDirective.KEEP,
+    new ScopedNameOfKind(MetadataKind.PROPERTY, MetadataScope.USER, "z"),
+    MetadataDirective.PRESERVE);
+
+  private final Metadata testMetadata = new Metadata(
+    ImmutableSet.of(new ScopedName(MetadataScope.SYSTEM, "a"),
+                    new ScopedName(MetadataScope.SYSTEM, "b"),
+                    new ScopedName(MetadataScope.USER, "c")),
+    ImmutableMap.of(new ScopedName(MetadataScope.SYSTEM, "x"), "1",
+                    new ScopedName(MetadataScope.SYSTEM, "y"), "2",
+                    new ScopedName(MetadataScope.USER, "z"), "4"));
+  @Test
+  public void testCreate() throws Exception {
+    final MetadataEntity createEntity = MetadataEntity.builder().append("create", "test").build();
+    createMetadataMutation(new MetadataMutation.Create(createEntity, testMetadata, CREATE_DIRECTIVES));
+
+    Assert.assertEquals(testMetadata.getProperties(MetadataScope.SYSTEM),
+                        getMetadataProperties(createEntity, MetadataScope.SYSTEM));
+    Assert.assertEquals(testMetadata.getTags(MetadataScope.SYSTEM),
+                        getMetadataTags(createEntity, MetadataScope.SYSTEM));
+    Assert.assertEquals(testMetadata.getProperties(MetadataScope.USER),
+                        getMetadataProperties(createEntity, MetadataScope.USER));
+    Assert.assertEquals(testMetadata.getTags(MetadataScope.USER),
+                        getMetadataTags(createEntity, MetadataScope.USER));
+
+    // confirm that KEEP/PRESERVE work by trying to re-create the entity
+    final Metadata recreate =
+      new Metadata(Collections.EMPTY_SET,
+                   ImmutableMap.of(new ScopedName(MetadataScope.SYSTEM, "x"), "10",
+                                   new ScopedName(MetadataScope.SYSTEM, "y"), "20",
+                                   new ScopedName(MetadataScope.USER, "z"), "40"));
+
+    createMetadataMutation(new MetadataMutation.Create(createEntity, recreate, CREATE_DIRECTIVES));
+    Assert.assertEquals(recreate.getProperties(MetadataScope.SYSTEM),
+                        getMetadataProperties(createEntity, MetadataScope.SYSTEM));
+    Assert.assertEquals(recreate.getTags(MetadataScope.SYSTEM),
+                        getMetadataTags(createEntity, MetadataScope.SYSTEM));
+    Assert.assertEquals(testMetadata.getProperties(MetadataScope.USER),
+                        getMetadataProperties(createEntity, MetadataScope.USER));
+    Assert.assertEquals(testMetadata.getTags(MetadataScope.USER),
+                        getMetadataTags(createEntity, MetadataScope.USER));
+  }
+
+  @Test
+  public void testRemove() throws Exception {
+    final MetadataEntity removeEntity = MetadataEntity.builder().append("remove", "test").build();
+    createMetadataMutation(new MetadataMutation.Create(removeEntity, testMetadata, CREATE_DIRECTIVES));
+
```

Comment:
Please assert that the create was successful.

Reply:
Can you do a remove that only removes specific tags/props? And then validate that only those were removed.